Feature/compile server jwt auth 리뷰 부탁드립니다.

kangsangmoon opened the pull request:
## #️⃣연관된 이슈

#37 

## 📝작업 내용

1. 컴파일 서버 측에 인증 기능을 추가
    -JWT 토큰을 데이터베이스에서 일치하는 지 검사
2. js, c++ 컴파일 언어 추가

kangsangmoon committed:
update(yml) :  datasource의 url 수정

kangsangmoon committed:
update(Dockerfile) :  JAR 파일을 ARG 변수로 변경

- ADD에서 COPY로 변경
- JAR 파일 경로를 ARG로 설정

kangsangmoon committed:
update(ResponseMessage) : INVALID_TOKEN_EXCEPTION 추가

- 토큰이 유효하지 않을 때 반환할 메시지 추가

kangsangmoon committed:
update(CompileApiController) : jwt 인증 추가 및 DTO 변경

- JWT 토큰 유효성 검사 추가
- Authorization 헤더로 인증 처리
- CompileRequest로 요청 DTO 변경

kangsangmoon committed:
update(CompileService) : 언어 및 임시 파일 삭제 처리 추가

- C++, JavaScript 언어 추가
- 코드 실행 및 컴파일 로직 간소화
- 스트림 처리 로직을 readStream 메서드로 통합
- 컴파일 후 남아있는 임시파일 삭제 처리

kangsangmoon committed:
update(compile) : 언어 선택 추가

kangsangmoon committed:
update(compile) : 코드 수정

kangsangmoon committed:
update(compile) : utf-8 인코딩

- 자바와 파이썬 인코딩 사용

kangsangmoon committed:
update(compile) : js 코드 수정

kangsangmoon committed:
chore(RestAPIController): RestAPI 컨트롤러들을 rest 패키지로 이동함

kangsangmoon committed:
update(compile) : hasAnyRole의 유저와 관리자 역할 수정

kangsangmoon committed:
Merge branch 'refs/heads/develop' into Feature/Compile_Server_JWT_Auth

kangsangmoon committed:
refactor(compile) : 불필요한 인증 어노테이션 삭제

Duskafka commented on src/main/java/com/example/project/compile/service/CompileService.java:
```diff
@@ -29,47 +33,38 @@ public String compileAndRun(String language, String code) throws IOException, In
                 case "java" -> compileJava(filePath);
                 case "python" -> compilePython(filePath);
+                case "javascript", "js" -> runJavaScript(filePath);
                 default -> throw new IllegalArgumentException(ErrorMessage.UNSUPPORTED_LANGUAGE.getMessage());
```
enum을 사용합시다

Duskafka commented on src/main/java/com/example/project/compile/service/CompileService.java:
클래스가 너무 큽니다. 역할을 나눠주세요

Duskafka commented on src/main/resources/static/js/Compile.js:
```diff
@@ -0,0 +1,25 @@
+    const code = document.getElementById("code").value;
+    const outputElement = document.getElementById("output");
+
+    try {
+        const response = await fetch('/api/compile', {
+            method: 'POST',
+            headers: { 'Content-Type': 'application/json; charset=UTF-8' },
+            body: JSON.stringify({ language, code })
+        });
```
토큰 인증 기능 추가해주세요

kangsangmoon committed:
Merge branch 'develop' into Feature/Compile_Server_JWT_Auth

kangsangmoon committed:
Merge branch 'develop' into Feature/Compile_Server_JWT_Auth

# Conflicts:
#	src/main/java/com/example/project/common/dto/ResponseMessage.java
#	src/main/java/com/example/project/compile/service/CompileService.java
#	src/main/resources/static/js/complie.js
#	src/main/resources/templates/compile/compile.html

kangsangmoon committed:
refactor(controller) : 분할 된 서비스에 따라 컨트롤러 수정

kangsangmoon committed:
refactor(controller) : 역할에 따라 코드 분할

kangsangmoon committed:
delete(js) : 중복된 js 삭제

kangsangmoon committed:
update(js) : 토큰 저장 추가

Duskafka commented on src/main/java/com/example/project/compile/controller/rest/CompileApiController.java:
```diff
@@ -1,36 +1,49 @@
-        } catch (Exception e) {
-            return ResponseDto.toResponseEntity(ResponseMessage.GENERAL_COMPILE_ERROR, e.getMessage());
-        }
+    public ResponseEntity<?> compileCode(@RequestHeader("code-for-code-auth") String token,
```
HttpServletRequest만 받아오면 토큰 값을 따로 받아오지 않아도 됩니다

Duskafka commented on src/main/java/com/example/project/compile/domain/Language.java:
클래스 명을 CompileLanguage로 변경하는 것이 좋을 것 같습니다

Duskafka commented on src/main/java/com/example/project/compile/domain/Language.java:
```diff
@@ -0,0 +1,38 @@
+        return compileCommand;
+    }
+
+    public static Language fromString(String language) {
```
메소드명에 예약어를 사용하지 말아주세요

Duskafka commented on src/main/java/com/example/project/compile/service/FileService.java:
```diff
@@ -0,0 +1,36 @@
+@Service
+public class FileService {
+
+    private static final String CODE_DIR = "/home/ubuntu/code/";
```
yml 에 설정한 값을 사용해주세요.
그리고 static을 사용할 필요는 없어보입니다.

kangsangmoon committed:
refactor(CompileLanguage) : 메서드명 및 클래스명 변경

- 메서드명 fromString에서 getByLanguageName으로 변경
- 클래스명 Language에서 CompileLanguage로 수정

kangsangmoon committed:
refactor(CompileController) : HttpServletRequest로 토큰 가져오기

- HttpServletRequest로 토큰값을 가져와서 인증이 되면 컴파일하게 수정

kangsangmoon committed:
refactor(FileService) : FileService 개선 및 코드 디렉터리 설정 변경

- @Value 어노테이션을 사용하여 외부 설정으로 주입
- Java 파일 생성 시 클래스 이름을 기반으로 파일 이름을 생성
- Java 코드로부터 클래스 이름을 추출하는 extractClassName 메서드 추가
- Files.writeString()를 사용하여 간결화

kangsangmoon committed:
refactor(CompileService) : Language 처리 방식 변경

- Java 코드의 클래스 이름 추출 로직을 FileService로 이동
- Language에서 CompileLanguage로 변경에 따른 메서드 호출 수정 (fromString -> getByLanguageName)

kangsangmoon committed:
refactor(Js) : Signup의 post 경로 수정

- jwt 삭제

Duskafka commented on src/main/java/com/example/project/compile/controller/rest/CompileApiController.java:
```diff
@@ -1,36 +1,48 @@
+    public ResponseEntity<?> compileCode(@RequestBody CompileRequest request, HttpServletRequest httpServletRequest) throws IOException {
+
+        if (authTokenService.isValidateToken(HeaderUtil.resolveToken(httpServletRequest))) {
+            try {
+                String result = compileService.compileAndRun(request.getLanguage(), request.getCode());
```
컴파일하는 서버에서도 JWT 토큰을 가지고 유효한 접근인지 아닌지 판단해서 컴파일할 수 있도록 해주세요

Duskafka commented on src/main/java/com/example/project/compile/service/FileService.java:
```diff
@@ -0,0 +1,55 @@
+@Service
+public class FileService {
+
+    private final String code_dir;
```
변수명은 카멜케이스를 지켜서 작성해주세요

kangsangmoon committed:
update(Compile) : JWT 토큰 재검사

- CompileService의 compileAndRun 메서드에 토큰이 일치하는 지 재검사 후 일치하는 경우에만 컴파일 하도록 수정

kangsangmoon committed:
update(Compile) : 변수명 카멜케이스에 맞게 수정

-

Duskafka committed:
refactor(Compile): 컴파일 서비스 코드 리펙토링 및 수정할 점 주석 추가

Duskafka committed:
refactor(Lombok): 애노테이션 추가해서 코드 간소화

kangsangmoon committed:
update(Compile) : 사용하지 않는 response 메시지 삭제

kangsangmoon committed:
update(Compile) : 유효하지 않은 경로와 접근 권한이 없을 때의 예외 추가

kangsangmoon committed:
refactor(Compile) : 언어 이름 배열 추가 및 switch 문 제거

- 언어에 대해 언어 이름 배열 추가
- 반복문을 사용하여 언어 검색

kangsangmoon committed:
refactor(Compile) : compileAndRun 메서드에 codeDir 추가 및 예외 처리

- compileAndRun 메서드 호출 시 codeDir 매개변수 추가
- 예외 처리 블록 제거하여 코드 간소화

kangsangmoon committed:
refactor(Compile) : readStream 메서드의 예외 처리 추가

kangsangmoon committed:
refactor(Compile) : 코드 디렉토리 매개변수 추가 및 코드 실행 로직 단순화

- 파일 삭제 시 Files.deleteIfExists 사용
- executeCode 메서드에서 언어별 실행 로직 제거
- compileAndRun 메서드에 codeDir 매개변수 추가

kangsangmoon committed:
refactor(Compile) : 코드 디렉토리 매개변수화 및 예외 처리

- createCodeFile 메서드에서 codeDir을 매개변수로 받아 필드 제거
- 파일 경로 생성 시 InvalidPathException 및 SecurityException 처리
- UUID를 사용하여 고유한 파일 이름 생성으로 파일 충돌 방지

kangsangmoon committed:
refactor(Compile) : static 제거

kangsangmoon committed:
refactor(Compile) : 컴파일 초기 언어값 java로 고정
